From bd8ccb7e0eefaac80028b4ffd16d09b91f3fe5f2 Mon Sep 17 00:00:00 2001
Message-ID: <bd8ccb7e0eefaac80028b4ffd16d09b91f3fe5f2.1746833337.git.sam@gentoo.org>
From: Filip Kastl <fkastl@suse.cz>
Date: Fri, 9 May 2025 16:04:25 +0200
Subject: [PATCH] gimple: Don't assert that switch has nondefault cases during
 lowering [PR120080]

I have mistakenly assumed that switch lowering cannot encounter a switch
with zero clusters.  This patch removes the relevant assert and instead
gives up bit-test lowering when this happens.

	PR tree-optimization/120080

gcc/ChangeLog:

	* tree-switch-conversion.cc (bit_test_cluster::find_bit_tests):
	Replace assert with return.

Signed-off-by: Filip Kastl <fkastl@suse.cz>
---
 gcc/tree-switch-conversion.cc | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/gcc/tree-switch-conversion.cc b/gcc/tree-switch-conversion.cc
index dea217a01efb..bd4de966892c 100644
--- a/gcc/tree-switch-conversion.cc
+++ b/gcc/tree-switch-conversion.cc
@@ -1793,12 +1793,14 @@ bit_test_cluster::find_bit_tests (vec<cluster *> &clusters, int max_c)
      end up with as few clusters as possible.  */
 
   unsigned l = clusters.length ();
-  auto_vec<min_cluster_item> min;
-  min.reserve (l + 1);
 
-  gcc_checking_assert (l > 0);
+  if (l == 0)
+    return clusters.copy ();
   gcc_checking_assert (l <= INT_MAX);
 
+  auto_vec<min_cluster_item> min;
+  min.reserve (l + 1);
+
   int bits_in_word = GET_MODE_BITSIZE (word_mode);
 
   /* First phase: Compute the minimum number of clusters for each prefix of the

base-commit: a470433732e77ae29a717cf79049ceeea3cbe979
-- 
2.49.0

